edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs

```diff
@@ -143,11 +143,11 @@ static async Task<int> MainAsync(IConfigurationRoot configuration)
                         logger.LogError($"Error starting protocol heads: {ex.Message}");
                     }
 
+                    await CloseDbStoreProviderAsync(container);

```

Comment:
One thing we should discuss is whether doing backup before protocol heads start to close will lead to issues.

Reply:
Why don't you just catch the timeout exception, or as I see you can override what it should do in case of timeout, and your implementation would not throw and exception? In that case, in normal circumstances, it would do what it's been doing (no behavior change), and in case of timeout, it still would store (with the risk that half-working protocol heads cause problems with backup)

Reply:
That is a good alternative, however we should consider this additional context:

When a user does not explicitly provide EdgeHub configuration env var "protocolTimeoutSecs" as some low value, the protocol heads will stay waiting for shutdwon until iotedged loses patience waiting for SIGTERM to take effect, and finally forcefully issues SIGKILL. This results in no timeout in the code, but actually the container abruptly stopping progress and a new one being created (and obviously we skip the backup operation which is the issue).

So maybe now what we should do is to change the default value of protocolTimeoutSecs to be some value more compatible with the time difference between SIGTERM and SIGKILL (i.e. some low value). However this suffers from the issue that the default value would then be very low, and upon a scenario like cert renewal where no SIGTERM is even issued, it is probably better to not cap the protocolTimeoutInSecs to some low value as there is no time pressure for waiting to shutdown. 

The motivation for my initial change was to sidestep making additional changes to address these points, as long as we are ok issuing CloseDbStoreProviderAsync() before the protocol heads start shutting down. But maybe we should do something more inline with what you are thinking, but we would also have to cap protocolTimeoutInSecs to a value compatible with SIGTERM/SIGKILL scenario.

Another option is to dynamically set protocolTimeoutSecs to some low value only if the `cts` for shutdown is cancelled, but leave it to the high default value if cert renewal is cancelled. However there are also edge cases with this approach where cert renewal and shutdown happen close together and therefore it is difficult to avoid an early SIGKILL.

We should also consider the fact that this change was intended to be something very straightforward to fix a scenario we don't really care about. So if this grows too much in complexity we may want to abandon making the change altogether.

Thoughts? My vote is to make the simple change to decrease default value of protocolTimeoutSecs, which will incur a more abrupt than necessary protocol head close on cert renewal scenarios. But with this approach a restart is always the same as EH getting restarted through `iotedge restart edgeHub` so it is conceptually easier to think about as if this scenario works everything will work.